### ISSUE TITLE
PCAP-over-IP client

### ISSUE THREAD

**Issue by erik4711** [zeek/zeek#2348]
Implement a PCAP-over-IP client, so that packets can be read from a TCP socket as an alternative to reading a pcap file or sniffing an interface.

PCAP-over-IP can be used to live-stream captured packets from a remote sniffer to zeek, but the most useful use-case is probably in order to read decrypted HTTPS traffic from a TLS proxy without having to replay it to a dummy network interface. There are several downsides of replaying packets to an interface, such as requiring root privs and risk of packet loss, so reading them from a TCP socket instead is preferable.

It would be nice if zeek would support the [same TCP socket interface syntax as Wireshark/tshark](https://wiki.wireshark.org/CaptureSetup/Pipes.md#tcp-socket).

`zeek -i TCP@192.168.1.2:57012`

Or if the TCP socket could be specified with -r:

`zeek -r TCP@192.168.1.2:57012`

Another option would be to add a custom option specifically for PCAP-over-IP:

`zeek --pcapoverip 192.168.1.2:57012`

**Comment by JustinAzoff** [comment]
This should be able to be implemented as a PktSrc plugin.  It would likely never support the same syntax as wireshark, but it could be close.  If the plugin was named pcapovertcp you'd run zeek like

    zeek -i pcapovertcp::192.168.1.2:57012

which would run the `pcapovertcp` plugin and pass `192.168.1.2:57012` to the constructor

A PktSrc plugin is basically two functions `Open`, and `ExtractNextPacket` and implementing this wouldn't be too much code.

**Comment by awelzel** [comment]
(mostly food for thought) reading the wireshark docs:

> A TCP stream is treated as like data from other pipes and the same restrictions apply. On each new connection the TCP server must send the header blocks as specified by libpcap or pcapng before any packet captures. 

Maybe the following gets close for testing? It may behave different from a "live packet source" due to the `-r` though, so maybe not appropriate for the use-case.
```
$ socat tcp:192.168.1.2:57012 - | zeek -C -r -
```

**Comment by timwoj** [comment]
I kinda like @JustinAzoff's suggestion of using a PktSrc here. That's almost exactly the point of a packet source. It would just need to open the socket and have packets available during the call to `ExtractNextPacket`. For that matter, you could  hook the socket itself into Zeek's kqueue and let it notify you when packets are available.

**Comment by ckreibich** [comment]
@erik4711 it sounds like you have an example of such a TLS proxy setup — could you share more details?

(Also, technically this'd be pcap-over-TCP, not pcap-over-IP.)

**Comment by erik4711** [comment]
> @erik4711 it sounds like you have an example of such a TLS proxy setup — could you share more details?

You mean this?
https://netresec.com/?b=228fddf
> 
> (Also, technically this'd be pcap-over-TCP, not pcap-over-IP.)

Yeah, it's a bit of a mystery why it's referred to as PCAP-over-IP rather than PCAP-over-TCP. See the FAQ in link above for more details.

**Comment by ckreibich** [comment]
>  https://netresec.com/?b=228fddf

Yeah, thank you — there's useful detail in there: the connection might go in the other direction (so Zeek would have to listen to support that), and the connection might use TLS. I like the PktSrc approach architecturally, but this means there's more to full support. Is there anything in Arne's `socat` suggestion that doesn't address what you need?

**Comment by erik4711** [comment]
> Yeah, thank you — there's useful detail in there: the connection might go in the other direction (so Zeek would have to listen to support that),

In most deployments the packet source is running the TCP listener. It would be great if Zeek additionally supported listening for incoming PCAP-over-IP connections (like Arkime does when started with `pcapReadMethod=pcap-over-ip-server`), but that would be more of a nice to have.

> and the connection might use TLS.

Most deployments run PCAP-over-IP over localhost to share packets between applications, such as from a TLS proxy to an IDS. In those cases TLS isn't really needed. Also adding TLS comes with additional complexity, such as deciding which certs should be trusted etc. My recommendation would therefore be to not require TLS support at this moment.

> I like the PktSrc approach architecturally, but this means there's more to full support.

From what I understand, the PktSrc approach seems to be enough to support PCAP-over-IP. 

> Is there anything in Arne's `socat` suggestion that doesn't address what you need?

Technically piping output from socat/netcat/ncat into STDIN of zeek does the trick. But from what I understand it's not really a desirable solution for those who do "proper" zeek deployments, which might involve running zeek with the Supervisor Framework or specifying multiple monitor interfaces in a  configuration file. I'm no zeek expert though, so I might be wrong here.

**Comment by erik4711** [comment]
> seems like we should support this, but it's not clear whether it should go on the 5.1 pile still, and I was concerned that if you have a pressing need and `socat` doesn't cut it, you'd be left with no options.

Many installations currently use a `nc localhost 57012 | tcpreplay -` combo to feed packets from PCAP-over-IP into Zeek. That’s a tolerable solution for most, so I don’t think there’s any urgency to force a PCAP-over-IP PktSrc implementation into the next release.

**Comment by JustinAzoff** [comment]
Also.. this doesn't have to be included in zeek (and subject to zeek release schedules), Many PktSrc plugins are external projects, like af_packet, pf_ring, myricom, netmap.

**Comment by erik4711** [comment]
It would be fantastic if there was a PCAP-over-IP PktSrc plugin in Zeek though.

**Comment by ckreibich** [comment]
Yes, Justin's point doesn't preclude that. It'd be a one-liner in `zkg` to install it. Great point.

**Comment by awelzel** [comment]
> Many installations currently use a nc localhost 57012 | tcpreplay - combo to feed packets from PCAP-over-IP into Zeek. That’s a tolerable solution for most, 

May or may not matter. If used with `af_packet` this approach provides load-balancing and multiple workers. With a pcap-over-ip PktSrc you'd need to find other ways for running multiple workers or it may act surprising - though could be cool to have a pcap-over-ip server/proxy implement load-balancing over connecting clients ;-)

**Comment by ckreibich** [comment]
While on this topic, there's also the related (but different, and more complex) rpcap:

https://github.com/the-tcpdump-group/libpcap/blob/master/rpcap-protocol.h
https://www.winpcap.org/docs/docs_412/html/group__remote.html
https://www.extrahop.com/company/blog/2014/lean-and-mean-our-open-source-enhancements-to-rpcap/

**Comment by simeonmiteff** [comment]
> May or may not matter. If used with `af_packet` this approach provides load-balancing and multiple workers. With a pcap-over-ip PktSrc you'd need to find other ways for running multiple workers or it may act surprising - though could be cool to have a pcap-over-ip server/proxy implement load-balancing over connecting clients ;-)

:thinking: this is interesting to me.

Seems like an ordinary TCP load balancer (without application layer specific handling) will do if the aim is to sink multiple PCAP-over-TCP connections on multiple workers.

Is that the intent, or would you want something like a (possibly flow-aware) packet broker (i.e., packets -> [broker/proxy] => multiple PCAP-over-TCP connections => multiple workers)?

**Comment by simeonmiteff** [comment]
> While on this topic, there's also the related (but different, and more complex) rpcap:
> 
> https://github.com/the-tcpdump-group/libpcap/blob/master/rpcap-protocol.h https://www.winpcap.org/docs/docs_412/html/group__remote.html https://www.extrahop.com/company/blog/2014/lean-and-mean-our-open-source-enhancements-to-rpcap/

My 2c: this seems OK if you deliberately truncate packets to fit in the MTU (you can configure ERSPAN to do this) but IMHO trying to avoid fragmentation with the udpstr mode means partially (poorly) re-inventing TCP, so you'd be better off using PCAP-over-TCP _or_ ERSPAN.

**Comment by awelzel** [comment]
> 
> Seems like an ordinary TCP load balancer (without application layer specific handling) will do if the aim is to sink multiple PCAP-over-TCP connections on multiple workers.
> 
> Is that the intent, or would you want something like a (possibly flow-aware) packet broker (i.e., packets -> [broker/proxy] => multiple PCAP-over-TCP connections => multiple workers)?

Was more thinking that the pcap-over-ip server could be flow-aware to support load-balancing. E.g., if multiple Zeek instances request the same endpoint, they'd magically get the right subset of traffic (think magic af-packet load balancing).

Though magic isn't always great, so could also encode load-balancing info into the URL, so that the server side then does  `flow-hash % #streams == stream_num` based on parameters in the URL. Four Zeek workers could then request different URLs:

```
# http://<server-ip>/#streams/#stream_num
http://<server-ip>/4/0
http://<server-ip>/4/1
http://<server-ip>/4/2
http://<server-ip>/4/3

# Or as query params...
# http://<server-ip>?streams=4&stream_num=0
```
But this is mostly wondering about things that could be neat. It's not clear to me that a pcap-over-ip use-case would actually require load-balancing across workers. And the packet source implementation wouldn't need to worry about this either.

**Comment by erik4711** [comment]
Most PCAP-over-TCP services will use IEEE 802.3 Ethernet (aka DLT_EN10MB) as the link_type, but it's best to assign the link type value dynamically based on the PCAP header. The first 24 bytes of the data stream received through PCAP-over-TCP is a PCAP header, which defines the link_type at offset 20. 

You might also wanna take a look at the PCAP file format breakdown that I've published here:
https://www.netresec.com/?page=Blog&month=2022-10&post=What-is-a-PCAP-file

**Comment by emnahum** [comment]
Thanks, I've looked at that URL.  My code has no problem reading the pcap_file_header for the initial read off the socket, and the linktype is set accordingly.  See https://github.com/emnahum/zeek-pcapovertcp-plugin/blob/master/src/PcapOverTcp.cc#L119.

Where I'm stuck now is the pcap_pkthdr, where, after the byte swap, the len is 6684672 and the caplen is 16801280, and things go downhill from there.  If I don't swap (which I think I should?), the len is 26112 and the caplen is 6160385, which I think is better, but then the ts.tv_sec and tv_usec don't make sense.  Any suggestions?  I'm assuming the pcap_pkthdr is in network (big-endian) order.

I've been looking for examples of PCAP-over-TCP in sources like wireshark, libcap but have had no luck.  Any suggestions there?  Thanks.

**Comment by emnahum** [comment]
Actually, the more I look at it, the more it seems that the pcap_pkthdr is in little-endian order, which it was written with. I captured some data on a X86 Ubuntu 20.04 box, and then sent it through socat to talk to the pcap-over-tcp plugin.
client: cat ../test.pcap | socat - tcp-listen:57012
server: zeek -i pcapovertcp::127.0.0.1:57012

**Comment by emnahum** [comment]
Urrgh.  After looking at a hexdump of the pcap file, and looking at https://www.netresec.com/?page=Blog&month=2022-10&post=What-is-a-PCAP-file again, I notice that what is being carried is a pcap save file, not a regular pcap as seen by a user space program.  In turns out you need to read the header with a struct pcap_sf_pkthdr, not a struct pcap_pkthdr, which is 8 bytes smaller.  Now things are working much better!  

The URL at netresec.com does not make this obvious, saying that a "packet header" is 16 bytes.  But this is the pcap_sf_pkthdr, not the pcap_pkthdr that might seem more obvious at first glance (at least to me).   It was only after digging around in https://github.com/the-tcpdump-group/libpcap/blob/master/pcap-int.h#L422 that I figured out what was being sent was a savefile.  The URL should explicitly mention the pcap_sf_pkthdr, which is not exported by the libpcap files.

Should PCAP-over-TCP include pcapng, or is that another "protocol"?

**Comment by erik4711** [comment]
> Urrgh. After looking at a hexdump of the pcap file, and looking at https://www.netresec.com/?page=Blog&month=2022-10&post=What-is-a-PCAP-file again, I notice that what is being carried is a pcap save file, not a regular pcap as seen by a user space program. In turns out you need to read the header with a struct pcap_sf_pkthdr, not a struct pcap_pkthdr, which is 8 bytes smaller. Now things are working much better!

Good find!

> The URL at netresec.com does not make this obvious, saying that a "packet header" is 16 bytes. But this is the pcap_sf_pkthdr, not the pcap_pkthdr that might seem more obvious at first glance (at least to me). It was only after digging around in https://github.com/the-tcpdump-group/libpcap/blob/master/pcap-int.h#L422 that I figured out what was being sent was a savefile. The URL should explicitly mention the pcap_sf_pkthdr, which is not exported by the libpcap files.

The "What is a PCAP file?" blog post as well as Wireshark's "Libpcap File Format" wiki page both cover the data structure of PCAP files (i.e. a savefile), not the PCAP API, so I'm not seeing any need to explicitly state that the packet header is referring to [pcap_sf_pkthdr](https://github.com/the-tcpdump-group/libpcap/blob/22813e20e8f1f1a1c1a4cd95cc06cd1e422a9dc0/pcap-int.h#L422) rather than [pcap_pkthdr](https://github.com/the-tcpdump-group/libpcap/blob/22813e20e8f1f1a1c1a4cd95cc06cd1e422a9dc0/pcap/pcap.h#L293). But you do have a valid point in that it might not be clear that PCAP-over-TCP is using the PCAP file format as the data structure, rather than providing a PCAP API wrapped in TCP. I'm not sure where this should be documented though. Thoughts?

> Should PCAP-over-TCP include pcapng, or is that another "protocol"?

My two cents would be to only do PCAP and avoid PcapNG.

You also asked about `is_live` before. I would argue that it should be set to `true` since there is little (or no) use for streaming a previously captured PCAP file as PCAP-over-TCP. It would make more sense to read that file from disk instead.

**Comment by simeonmiteff** [comment]
On Tue, Apr 4, 2023, 19:16 erik4711 ***@***.***> wrote:

> Should PCAP-over-TCP include pcapng, or is that another "protocol"?
>
> My two cents would be to only do PCAP and avoid PcapNG..
>

My 2c Australian agrees - PCAPNG is way more complicated and taking
advantage of the extra features it has (while interesting) is probably out
of scope for this issue.

**Comment by emnahum** [comment]
Well, in the "what is a pcap file" you could add some links to the proper
data structure like you did above, and just mention that it's to a save
file.  There are links for things like the Ethernet type.

There's no *explicit* mention of a savefile in the URL.  So, to someone who
lives and breathes PCAP, it's obvious that's what you are referring to, but
as someone without that knowledge that came to the site, it wasn't clear.

PcapNG:  Agreed.  Way more complicated (per Simeon's comment).

is_live: I'm not so sure, but at the moment, I'm assuming that (e.g.,
calling the constructor with is_live==false results in an error).

**Comment by erik4711** [comment]
Thanks for the suggestion @emnahum! The [What is a PCAP file?](https://www.netresec.com/?page=Blog&month=2022-10&post=What-is-a-PCAP-file) blog post has now been updated to include the term "pcap savefile" and links have been added to the `pcap_file_header` and `pcap_sf_pkthdr` structs in libpcap's source code.

**Comment by emnahum** [comment]
Great, thank you!

**Comment by emnahum** [comment]
The plugin is ready (I think).  Please take a look at https://github.com/emnahum/zeek-pcapovertcp-plugin.

**Comment by erik4711** [comment]
That's great news! I have now updated the PCAP-over-IP blog post to include a link to the plugin.
